### PR TITLE
Multi turn servo

### LINKED ISSUE #1090: Add support for winch/multi-turn servo
[Winch servos](http://hitecrcd.com/products/servos/boat-servos/hs-785hb-3.5-turn-winch-servo/product) are basically normal servos with a multi-turn pot. Surprisingly, not many people know about them but they can be super handy for pan-tilt things.

To control them using the standard Servo arduino library, one needs to scale the values passws to the [`write()` call](http://forum.arduino.cc/index.php?topic=120884.msg910323#msg910323). 

I could see it added as a new `type` in the `Servo` component class.

### PR DESCRIPTION
**Pull request by dtex**
* Adds support for 90, 360 and multi-turn servos.
* Send servoWrites with values in microseconds for pulse width instead of degrees.
* Closes #1090 

This is for testing and cannot be merged until we confirm that all the IO plug-ins support passing position in degrees or pulse width in microseconds.

 - [ ] [imp-io/tyrion PR #8](https://github.com/rwaldron/tyrion/pull/8) [50hz, 600us - 2000us , 3% - 10%]
- [ ] [tessel-io PR #21](https://github.com/rwaldron/tessel-io/pull/21) [50hz, 600us - 2400us , 3% - 12%]
- [ ] [raspi-io/rasp-io-core Issue #4](https://github.com/nebrius/raspi-io-core/issues/4) [50hz, 1000us - 2000us, 5% - 10%]
- [ ] [galileo-io/edison-io](https://github.com/rwaldron/galileo-io/blob/b8f5f9ed9c2866f5557317fb1436a9c0a21e4f08/lib/index.js#L566-L576) [125.5hz, 600us - 2400us, 7.5% - 30.12%]
- [ ] [particle-io/voodoospark](https://github.com/voodootikigod/voodoospark/blob/master/firmware/voodoospark.cpp#L1009-L1019) [50hz, 1000us - 2000us, 5% - 10%]
- [ ] [nino-io](https://github.com/rwaldron/nino-io/blob/a676fda7331b988f9f0685218a7580b06d1988ae/lib/nino-io.js#L472-L482) [50z, 1000us - 2000us, 5% - 10%]
 - [ ] [beaglebone-io](https://github.com/julianduque/beaglebone-io) calls linux-io [50hz, 600us - 2400us, 3% - 12%]
 - [ ] [linux-io](https://github.com/fivdi/linux-io/blob/v0.8.0/lib/linux-io.js#L229-L247) [50hz, 600us - 2400us, 3% - 12%]
 - [x] [firmata.js]() [50hz, 544us - 2400us, 2.72% - 12%] Confirmed w/ scope
 - [x] [blend-micro-io](https://github.com/noopkat/blend-micro-io) servoWrite not implemented
 - [x] [board-io](https://github.com/achingbrain/board-io/blob/cf1c608500261c74bbf1d7a05aa3bff58dd70a62/lib/board-io.js#L136-L138) Not an issue
 - [x] [chip-io](https://github.com/sandeepmistry/node-chip-io)  servoWrite not implemented
 - [x] [bean-io](https://github.com/monteslu/bean-io/) servoWrite not implemented
 - [x] [pcduino-io](https://github.com/rwaldron/pcduino-io/blob/811c050c399724c2f8fbc41aeafe2b3d955c015d/lib/pcduino-io.js#L263-L265) servoWrite not implemented
 - [x] [pinnoccio-io](https://github.com/soldair/pinoccio-io/blob/4c76cbd1dacf0648803ab4b97b8245a1c791bb3d/index.js#L312) servoWrite not implemented. Device no longer manufactured.
 - [x] [remote-io](https://github.com/monteslu/remote-io) Not an issue

### CODE REVIEW

**Comment by dtex** [comment]
I think I'm going to need to get my hands on an oscilloscope so I can reconcile why some io plug-ins are using different PWM ranges.

File: README.md [hunk 1]
```diff
@@ -172,7 +172,11 @@ To get you up and running quickly, we provide a variety of examples for using ea
 
 To interactively navigate the examples, visit the [Johnny-Five examples](http://johnny-five.io/examples/) page on the official website. If you want to link directly to the examples in this repo, you can use one of the following links.
 
+<<<<<<< HEAD
 **There are presently 357 example programs with code and diagrams!**
+=======
+**There are presently 355 example programs with code and diagrams!**
+>>>>>>> servoWrite with microseconds instead of degrees.
```
Comment (rwaldron) [inline README.md]: Conflict markers

File: lib/servo.js [hunk 2]
```diff
@@ -97,7 +105,9 @@ function Servo(opts) {
     this, opts = Board.Options(opts)
   );
 
-  this.range = opts.range || [0, 180];
+  this.deviceRange = opts.deviceRange || [0, 180];
```
Comment (rwaldron) [inline lib/servo.js]: The value is always an array of degrees? What about `degreeRange`?

File: docs/servo-multi-turn.md [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+
+```javascript
+var five = require("johnny-five");
+var board = new five.Board();
+
+board.on("ready", function() {
+  var servo = new five.Servo({
+    pin: 10,
+    pwmRange: [600, 2370], // Cheapo servo has limited PWM range
+    deviceRange: [0, 2430] // This multi-turn servo can turn 2430 degrees or 6.75 revolutions
```
Comment (rwaldron) [inline docs/servo-multi-turn.md]: nit: put comments above the line (throughout), like:

```js
// Cheapo servo has limited PWM range
pwmRange: [600, 2370],
 // This multi-turn servo can turn 2430 degrees or 6.75 revolutions
deviceRange: [0, 2430]
```

The reason I ask for this is to improve johnny-five.io readiability on non-desktop/laptop devices

**Comment by dtex** [comment]
I'll make these changes, but there are bigger problems here. The way things currently are, the micro-controller can only know the servo's pwm range and Johnny-Five (and firmata.js) can only know the device range. We need to know both in order to scale correctly.

I can fix this by setting the pwmRange from Johnny-Five (requires servoConfig in all the io-plugins) but the current PWM ranges are inconsistent and I don't know why. Servo's are supposed to be manufactured to a standard range of 1000 - 2000 microseconds at 50hz, but we have ranges from 544 - 2400 microseconds down to 1000 - 2000. 

My suspicion is that authors just went with whatever values seemed to work with the servo(s) they had on hand. The Arduino default of 544 - 2400 is so wide as to encompass all possible servos, but it could also cause some to appear to not work if the requested value is beyond the device's functional range.

To confirm this I need to put the different micro-controllers on an oscilloscope and see what they are actually outputting. Based on that we need standardize the servo pwmRanges.

I don't currently have access to an oscilloscope, but they have several at the maker space I go to. To get access I have to take an electronics class, so this Saturday I'm taking a six hour "Introduction to Arduino" so I can get access to the oscilloscopes.

File: lib/servo.js (hunk 2 above)
Comment (dtex) [reply lib/servo.js]: I don't have strong feelings either way. degreeRange works for me.

**Comment by dtex** [comment]
I've scoped some micro controllers and so far all are outputting exactly the PWM width we ask for so the variation in ranges is not purposeful. I propose that J5 sets the same custom PWM range on all micro controllers so that servo behavior is the same across all IO plugins.

I propose 600 - 2400us. This matches Tessel, Edison and is really close on Arduino.  I'm certain that the 1000 - 2000us ranges are insufficient for every servo I've ever worked with.

**Comment by fivdi** [comment]
@dtex beaglebone-io supports servoWrite. The BeagleBone specific code is [here](https://github.com/julianduque/beaglebone-io/blob/v2.2.0/lib/beaglebone.js#L81-L89). The Linux specifix code is [here](https://github.com/fivdi/linux-io/blob/v0.8.0/lib/linux-io.js#L229-L247).

**Comment by dtex** [comment]
@fivdi Thanks! I missed the ```LinuxIO.call(...```

**Comment by dtex** [comment]
There are a lot of things that need to land in other repos before this PR is ready. I'm just going to close it for now to minimize clutter.

**Comment by dtex** [comment]
Not going to bother updating nino-io or galileo/edison so marking those as done.

**Pull request closed without merging**